File: app/src/main/java/uk/whitecrescent/waqti/android/customview/AutoScroller.java
package uk.whitecrescent.waqti.android.customview;

import android.content.Context;
import android.os.Handler;

class AutoScroller {
    enum AutoScrollMode {
        POSITION, COLUMN
    }

    enum ScrollDirection {
        UP, DOWN, LEFT, RIGHT
    }

    interface AutoScrollListener {
        void onAutoScrollPositionBy(int dx, int dy);

        void onAutoScrollColumnBy(int columns);
    }

    private static final int SCROLL_SPEED_DP = 8;
    private static final int AUTO_SCROLL_UPDATE_DELAY = 12;
    private static final int COLUMN_SCROLL_UPDATE_DELAY = 1000;

    private Handler mHandler = new Handler();
    private AutoScrollListener mListener;
    private boolean mIsAutoScrolling;
    private int mScrollSpeed;
    private long mLastScrollTime;
    private AutoScrollMode mAutoScrollMode = AutoScrollMode.POSITION;

    AutoScroller(Context context, AutoScrollListener listener) {
        mListener = listener;
        mScrollSpeed = (int) (context.getResources().getDisplayMetrics().density * SCROLL_SPEED_DP);
    }

    void setAutoScrollMode(AutoScrollMode autoScrollMode) {
        mAutoScrollMode = autoScrollMode;
    }

    boolean isAutoScrolling() {
        return mIsAutoScrolling;
    }

    void stopAutoScroll() {
        mIsAutoScrolling = false;
    }

    void startAutoScroll(ScrollDirection direction) {
        switch (direction) {
            case UP:
                startAutoScrollPositionBy(0, mScrollSpeed);
                break;
            case DOWN:
                startAutoScrollPositionBy(0, -mScrollSpeed);
                break;
            case LEFT:
                if (mAutoScrollMode == AutoScrollMode.POSITION) {
                    startAutoScrollPositionBy(mScrollSpeed, 0);
                } else {
                    startAutoScrollColumnBy(1);
                }
                break;
            case RIGHT:
                if (mAutoScrollMode == AutoScrollMode.POSITION) {
                    startAutoScrollPositionBy(-mScrollSpeed, 0);
                } else {
                    startAutoScrollColumnBy(-1);
                }
                break;
        }
    }

    private void startAutoScrollPositionBy(int dx, int dy) {
        if (!mIsAutoScrolling) {
            mIsAutoScrolling = true;
            autoScrollPositionBy(dx, dy);
        }
    }

    private void autoScrollPositionBy(final int dx, final int dy) {
        if (mIsAutoScrolling) {
            mListener.onAutoScrollPositionBy(dx, dy);
            mHandler.postDelayed(new Runnable() {
                @Override
                public void run() {
                    autoScrollPositionBy(dx, dy);
                }
            }, AUTO_SCROLL_UPDATE_DELAY);
        }
    }

    private void startAutoScrollColumnBy(int columns) {
        if (!mIsAutoScrolling) {
            mIsAutoScrolling = true;
            autoScrollColumnBy(columns);
        }
    }

    private void autoScrollColumnBy(final int columns) {
        if (mIsAutoScrolling) {
            if (System.currentTimeMillis() - mLastScrollTime > COLUMN_SCROLL_UPDATE_DELAY) {
                mListener.onAutoScrollColumnBy(columns);
                mLastScrollTime = System.currentTimeMillis();
            } else {
                mListener.onAutoScrollColumnBy(0);
            }

            mHandler.postDelayed(new Runnable() {
                @Override
                public void run() {
                    autoScrollColumnBy(columns);
                }
            }, AUTO_SCROLL_UPDATE_DELAY);
        }
    }
}

File: app/src/main/java/uk/whitecrescent/waqti/android/customview/AutoScroller.kt
package uk.whitecrescent.waqti.android.customview

import android.content.Context
import android.os.Handler

/*
 * This is just a Self Scrolling Entity, no views here just fancy Math that does the self
 * scrolling based on parameters and different types and directions of scrolling
 */
class AutoScroller(context: Context, private val listener: AutoScrollListener) {

    private val scrollSpeedDP = 8
    private val autoScrollUpdateDelay = 12L
    private val columnScrollUpdateDelay = 1000L

    private val handler = Handler()
    var isAutoScrolling = false
        private set

    private val scrollSpeed = (context.resources.displayMetrics.density * scrollSpeedDP).toInt()
    private var lastScrollTime = 0L
    var autoScrollMode = AutoScrollMode.POSITION

    fun startAutoScroll(direction: ScrollDirection) {
        when (direction) {
            ScrollDirection.UP -> startAutoScrollPositionBy(0, scrollSpeed)
            ScrollDirection.DOWN -> startAutoScrollPositionBy(0, -scrollSpeed)
            ScrollDirection.LEFT -> {
                if (autoScrollMode == AutoScrollMode.POSITION) startAutoScrollPositionBy(scrollSpeed, 0)
                else startAutoScrollColumnBy(1)
            }
            ScrollDirection.RIGHT -> {
                if (autoScrollMode == AutoScrollMode.POSITION) startAutoScrollPositionBy(-scrollSpeed, 0)
                else startAutoScrollColumnBy(-1)
            }
        }
    }

    fun stopAutoScroll() {
        isAutoScrolling = false
    }

    private fun startAutoScrollPositionBy(dx: Int, dy: Int) {
        if (!isAutoScrolling) {
            isAutoScrolling = true
            autoScrollPositionBy(dx, dy)
        }
    }

    private fun autoScrollPositionBy(dx: Int, dy: Int) {
        if (isAutoScrolling) {
            listener.onAutoScrollPositionBy(dx, dy)
            handler.postDelayed({
                autoScrollPositionBy(dx, dy)
            }, autoScrollUpdateDelay)
        }
    }

    private fun startAutoScrollColumnBy(columns: Int) {
        if (!isAutoScrolling) {
            isAutoScrolling = true
            autoScrollColumnBy(columns)
        }
    }

    private fun autoScrollColumnBy(columns: Int) {
        if (isAutoScrolling) {
            if (System.currentTimeMillis() - lastScrollTime > columnScrollUpdateDelay) {
                listener.onAutoScrollColumnBy(columns)
                lastScrollTime = System.currentTimeMillis()
            } else listener.onAutoScrollColumnBy(0)

            handler.postDelayed({
                autoScrollColumnBy(columns)
            }, autoScrollUpdateDelay)
        }
    }

    enum class AutoScrollMode {
        POSITION, COLUMN
    }

    enum class ScrollDirection {
        UP, DOWN, LEFT, RIGHT
    }

    interface AutoScrollListener {
        fun onAutoScrollPositionBy(dx: Int, dy: Int)

        fun onAutoScrollColumnBy(columns: Int)
    }
}